### PR TITLE
Add Last Scan endpoint

### PR DESCRIPTION
**Pull request by meln1k**
# Description

# To-Dos

- [x] Add test coverage for new or updated functionality
- [x] Lint and test with `nox`

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://some.engineering/code-of-conduct).

### CODE REVIEW

File: fixbackend/app.py [hunk 1]
```diff
@@ -142,7 +145,21 @@ async def setup_teardown_application(app: FastAPI) -> AsyncIterator[None]:
...
+            redis=readwrite_redis, channel="cloud_accounts", publisher_name="cloud_account_service"
+        )
+        deps.add(
+            SN.cloud_account_service,
+            CloudAccountServiceImpl(
+                workspace_repo,
+                CloudAccountRepositoryImpl(session_maker),
+                cloud_accounts_redis_publisher,
+                domain_event_publisher,
+                JsonStoreImpl(session_maker),
```
Comment (aquamatthias) [inline fixbackend/app.py:160]: I wonder why there is *the* json store.
Don't we want to use different underlying tables based on use case?

File: fixbackend/cloud_accounts/models/__init__.py [hunk 2]
```diff
@@ -38,3 +40,16 @@ class CloudAccount:
     id: CloudAccountId
     workspace_id: WorkspaceId
     access: CloudAccess
+
+
+@frozen
+class LastScanAccountInfo:
+    aws_account_id: str
```
Comment (aquamatthias) [inline fixbackend/cloud_accounts/models/__init__.py:47]: The value that is reported is an account id - not necessarily an AWS account id.

File: fixbackend/cloud_accounts/models/__init__.py [hunk 3]
```diff
@@ -38,3 +40,16 @@ class CloudAccount:
...
+    duration_seconds: int
```
Comment (aquamatthias) [inline fixbackend/cloud_accounts/models/__init__.py:48]: It is missing information about the last scanned time?

File: fixbackend/cloud_accounts/models/__init__.py [hunk 4]
```diff
@@ -38,3 +40,16 @@ class CloudAccount:
...
+    resources_scanned: int
+
+
+@frozen
+class LastScanInfo:
```
Comment (aquamatthias) [inline fixbackend/cloud_accounts/models/__init__.py:53]: can we add the time when the tenant collect has been triggered?

File: fixbackend/cloud_accounts/router.py [hunk 5]
```diff
@@ -35,6 +35,31 @@ async def delete_cloud_account(
...
         await service.delete_cloud_account(cloud_account_id, workspace.id)
 
+    @router.get("/{workspace_id}/cloud_accounts/last_scan")
+    async def last_scan(
+        workspace: UserWorkspaceDependency,
+        service: CloudAccountServiceDependency,
+    ) -> LastScanInfo:
+        last_scan = await service.last_scan(workspace.id)
+        if last_scan is None:
+            return LastScanInfo(
```
Comment (aquamatthias) [inline fixbackend/cloud_accounts/router.py:45]: consider moving this to a static method like `LastScanInfo.empty()`

File: fixbackend/cloud_accounts/service_impl.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    async def start(self) -> Any:
+        return await self.domain_event_listener.start()
+
+    async def stop(self) -> Any:
+        return await self.domain_event_listener.stop()
+
+    async def process_domain_event(self, message: Json, context: MessageContext) -> None:
+        match context.kind:
+            case TenantAccountsCollected.kind:
+                event = TenantAccountsCollected.from_json(message)
```
Comment (aquamatthias) [inline fixbackend/cloud_accounts/service_impl.py:75]: Like it!

File: fixbackend/dispatcher/dispatcher_service.py [hunk 7]
```diff
@@ -190,12 +206,30 @@ async def collect_job_finished(self, message: Json) -> None:
...
+            self._jobs_hash_key(workspace_id), job_id
+        )  # type: ignore
+        if cloud_account_id_bytes is None:
+            log.error(f"Could not find cloud account id for job id {job_id}")
+            return
+        cloud_account_id = cloud_account_id_bytes.decode()
+
+        account_progress_bytes: Optional[bytes] = await self.temp_store_redis.hget(
+            self._collect_progress_hash_key(workspace_id), cloud_account_id
+        )  # type: ignore
```
Comment (aquamatthias) [inline fixbackend/dispatcher/dispatcher_service.py:220]: redis should give you strings not bytes - or?

File: fixbackend/dispatcher/dispatcher_service.py [hunk 8]
```diff
@@ -221,11 +255,27 @@ async def compute_next_run(self, tenant: WorkspaceId, last_run: Optional[datetim
...
+            account_id = account_information.aws_account_id
+        else:
+            raise NotImplementedError("Unsupported account information type")
+        value = AccountCollectProgress(
+            cloud_account_id=cloud_account_id, account_id=account_id, started_at=utc()
+        ).to_json_str()
+        # store account_collect_progress
+        await self.temp_store_redis.hset(name=self._collect_progress_hash_key(workspace_id), key=str(cloud_account_id), value=value)  # type: ignore # noqa
+        # store job_id -> cloud_account_id mapping
+        await self.temp_store_redis.hset(name=self._jobs_hash_key(workspace_id), key=str(job_id), value=str(cloud_account_id))  # type: ignore # noqa
```
Comment (aquamatthias) [inline fixbackend/dispatcher/dispatcher_service.py:275]: why do we need two hash keys for one entry?

File: fixbackend/dispatcher/dispatcher_service.py (hunk 7 above)
Comment (meln1k) [reply fixbackend/dispatcher/dispatcher_service.py:220]: But it’s not.

**Review by aquamatthias — APPROVED** [review]
thanks!

**Review by aquamatthias — DISMISSED** [review]

**Review by aquamatthias — APPROVED** [review]

**Pull request merged**